### PR TITLE
Feat 344 slap

### LINKED ISSUE #344: Add Slap2 devices to rig schema

### LINKED ISSUE #346: Create slap2 parents session class in ophys session

### LINKED ISSUE #347: Create Branch imaging class in ophys session

### PR DESCRIPTION
**Pull request by saskiad**
closes #344
closes #346
closes #347

### CODE REVIEW

**Comment by saskiad** [comment]
@KasparP Trying to get these finished - we revamped the rig/session schema overall so now I'm adding in the slap2 stuff. I still have some questions about the devices - and what are stable parameters (rig schema) and what are variable parameters (session schema). For now, take a look at this and see if anything jumps out to you. I'll find some time when we can sit down and look at it when I'm back in town.

**Review by KasparP — APPROVED** [review]
All changes look good. We will work on translating metadata to this format in coming months, which will probably require further changes of this type.

File: src/aind_data_schema/device.py [hunk 1]
```diff
@@ -560,6 +568,44 @@ class Patch(Device):
...
+    line_shear: List[int] = Field(..., title="Line shear (pixels)")
+    line_shear_units: SizeUnit = Field(SizeUnit.PX, title="Line shear units")
+
+
+class PolygonalScanner(Device):
+    """Description of a Polygonal scanner"""
+
+    device_type: Literal["Polygonal Scanner"] = Field("Polygonal Scanner", const=True, readOnly=True)
+    speed: int = Field(..., title="Speed (rpm)")
+    speed_unit: str = Field("Rotations per minute", title="Speed unit")
```
Comment (dyf) [inline src/aind_data_schema/device.py:596]: should this be a unit in units.py?

File: src/aind_data_schema/device.py [hunk 2]
```diff
@@ -560,6 +568,44 @@ class Patch(Device):
...
+    number_faces: int = Field(..., title="Number of faces")
+
+
+class PockelsCell(Device):
+    """Description of a Pockels Cell"""
+
+    device_type: Literal["Pockels Cell"] = Field("Pockels Cell", const=True, readOnly=True)
+    polygonal_scanner: str = Field(..., title="Polygonal scanner", description="Must match name of Polygonal scanner")
+    on_time: Decimal = Field(..., title="On time (fraction of cycle)")
```
Comment (dyf) [inline src/aind_data_schema/device.py:605]: add "fraction of cycle" as unit?

File: src/aind_data_schema/session.py [hunk 3]
```diff
@@ -122,11 +122,30 @@ class Stack(AindModel):
     magnification: Optional[str] = Field(None, title="Magnification")
     fov_scale_factor: float = Field(..., title="FOV scale factor (um/pixel)")
     fov_scale_factor_unit: str = Field("um/pixel", title="FOV scale factor unit")
-    frame_rate: float = Field(..., title="Frame rate (Hz)")
+    frame_rate: Decimal = Field(..., title="Frame rate (Hz)")
     frame_rate_unit: FrequencyUnit = Field(FrequencyUnit.HZ, title="Frame rate unit")
     targeted_structure: Optional[str] = Field(None, title="Targeted structure")
 
 
+class SlapType(Enum):
```
Comment (dyf) [inline src/aind_data_schema/session.py:130]: `SlapSessionType`?

File: src/aind_data_schema/session.py [hunk 4]
```diff
@@ -238,6 +257,7 @@ class Stream(AindModel):
     detectors: Optional[List[Detector]] = Field(None, title="Detectors", unique_items=True)
     fiber_photometry_assemblies: Optional[List[FiberPhotometryAssembly]] = Field(None, title="Fiber photometry devices")
     ophys_fovs: Optional[List[FieldOfView]] = Field(None, title="Fields of view", unique_items=True)
+    slap_fovs: Optional[Slap] = Field(None, title="Slap2 field of view")
```
Comment (dyf) [inline src/aind_data_schema/session.py:260]: Can we rename `Slap` -> `SlapFieldOfView`? We have `Slap` as a modality and that would make the distinction clear.

**Review by dyf — CHANGES_REQUESTED** [review]
A few small suggestions.

File: src/aind_data_schema/device.py (hunk 1 above)
Comment (saskiad) [reply src/aind_data_schema/device.py:596]: I don't expect us to use it in other places, so I'm not sure it makes sense to do that. But we can if you think we should

File: src/aind_data_schema/device.py (hunk 2 above)
Comment (saskiad) [reply src/aind_data_schema/device.py:605]: it's technically unitless

Comment (saskiad) [reply src/aind_data_schema/device.py:605]: or I guess that could be the unit... like percent... okay I'll do that

**Review by dyf — APPROVED** [review]

**Pull request merged**